### DISCUSSION TITLE
Merging Maps in AggregatedMergeTree

### DISCUSSION THREAD

**Discussion by mdonkers** [ClickHouse/ClickHouse#50776] [Q&A] [answered]
Hi,

I'm trying to create a `AggregatedMergeTree` with a column with `Map` type. We're expecting very frequent updates with almost always the same data. But when receiving some differing key-value pairs, they should ideally get merged.

Currently I have:

```sql
CREATE TABLE IF NOT EXISTS aggregated_attributes
(
    `Name` String,
    `Attributes` SimpleAggregateFunction(maxMap, Map(String, String))
)
ENGINE = ReplicatedAggregatingMergeTree
PRIMARY KEY Name
ORDER BY Name
```

This will combine records, but won't merge the map keys or values. E.g.:

```sql
INSERT INTO aggregated_attributes VALUES('Foo', {'key1':'value1', 'key1':'value2'}),('Bar', {'key1':'value1','key2':'value2'})

SELECT *
FROM aggregated_attributes

Query id: ce682502-5e4f-44bc-adbf-dc1fff38ec7d

┌─Name─┬─Attributes────────────────────────┐
│ Bar  │ {'key1':'value1','key2':'value2'} │
│ Foo  │ {'key1':'value2'}                 │
└──────┴───────────────────────────────────┘

2 rows in set. Elapsed: 0.001 sec. 
```

With a different key it's even worse as it doesn't combine the keys but just takes the last Map:

```sql
INSERT INTO aggregated_attributes VALUES('Foo', {'key1':'value1', 'key3':'value3'}),('Bar', {'key1':'value1','key2':'value2'})

SELECT *
FROM aggregated_attributes
FINAL

Query id: 9c1eb485-fe3f-4357-9d2c-25ce8da2fd3d

┌─Name─┬─Attributes────────────────────────┐
│ Bar  │ {'key1':'value1','key2':'value2'} │
│ Foo  │ {'key1':'value2','key3':'value3'} │
└──────┴───────────────────────────────────┘
```

Is there any other function (that works with the `SimpleAggregateFunction`) that can correctly combine maps?
Or would that require using a Nested object instead?

Any ideas would be very welcome. Thx.

**Comment by den-crane** [comment]
???

>ORDER BY Name

Rows with different table KEY should not be merged. 'Foo' <> 'Bar'.


```
select maxMap( map ('key1','ABC','key2','value2', 'key1', 'ZZZ' ) );
┌─maxMap(map('key1', 'ABC', 'key2', 'value2', 'key1', 'ZZZ'))─┐
│ {'key1':'ZZZ','key2':'value2'}                              │
└─────────────────────────────────────────────────────────────┘
```
maxMap calculates max value of each map key

**Reply by mdonkers** [reply]
I see I made some mistake with that second insert, where I assumed three entries to be present in the map. But different keys actually get correctly merged...

See
```sql
INSERT INTO aggregated_attributes VALUES('Foo', {'key1':'value1', 'key2':'value2'}),('Bar', {'key1':'value1','key2':'value2'})

SELECT *
FROM aggregated_attributes
FINAL

Query id: 0a66c00d-cb8c-4a6d-acf9-3f18ebb24632

┌─Name─┬─Attributes────────────────────────────────────────┐
│ Bar  │ {'key1':'value1','key2':'value2'}                 │
│ Foo  │ {'key1':'value2','key2':'value2','key3':'value3'} │
└──────┴───────────────────────────────────────────────────┘
```

Which shows `key1`, `key2` and `key3` from separate inserts got merged into the same map.

What I'd like to have though (if possible) is `key1: [value1, value2]`
Currently `value2` wins because it's larger than `value1`, which makes sense for `maxMap`.

So I'm wondering if there is any more advanced merging possible for Maps, for the `AggregatedMergeTree`

**Reply by den-crane** [reply, marked as answer]
It's not related to AggregatedMergeTree.
There is no such aggregate function which can combine Map values into the array.

Something like this is possible with groupArray

```
CREATE TABLE IF NOT EXISTS aggregated_attributes(  
   Name String,
   Key String,
   Value SimpleAggregateFunction(groupArrayArray, Array(String))
)ENGINE = AggregatingMergeTree
ORDER BY (Name, Key);

insert into aggregated_attributes values ('foo', 'key1', ['X']);
insert into aggregated_attributes values ('foo', 'key1', ['Y']);

select * from aggregated_attributes final;
┌─Name─┬─Key──┬─Value─────┐
│ foo  │ key1 │ ['X','Y'] │
└──────┴──────┴───────────┘
```
And convert into required form in MatView

**Reply by mdonkers** [reply]
Thanks @den-crane , will try some things with different types than `Map`.
That should work better looking at your example.

**Comment by tbragin** [comment]
Hi @mdonkers! Can't wait to see what exciting stuff you are working on next! ❤️
